extensions/okhttp/src/main/java/com/google/android/exoplayer2/ext/okhttp/OkHttpDataSource.java

```diff
@@ -304,14 +304,14 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
 
     // Check for a valid response code.
     if (!response.isSuccessful()) {
+      Map<String, List<String>> headers = response.headers().toMultimap();
       byte[] errorResponseBody;
       try {
         errorResponseBody = Util.toByteArray(Assertions.checkNotNull(responseByteStream));
       } catch (IOException e) {
-        throw new HttpDataSourceException(
-            "Error reading non-2xx response body", e, dataSpec, HttpDataSourceException.TYPE_OPEN);
+        throw new InvalidResponseCodeException(
```

Comment:
Isn't it better just to replace this with a `errorResponseBody = Util.EMPTY_BYTE_ARRAY` and then fall through, so that the lines below take care of throwing what you want? I think that should just work. Note that you need to do it this way for the special handling of `416` to work as well.

Note that it will not be possible to disambiguate between "no response body" and "failed to read response body" in this case. If you think it's important to be able to disambiguate then you can use `errorResponseBody = null` instead. I'm not sure it's worth it though?